### PR TITLE
`react`: `Children.map`: replace overload with conditional type

### PR DESCRIPTION
**Pull request by OliverJAsh**
Adds to #41760 

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/react/test/index.ts [hunk 1]
```diff
@@ -533,6 +534,8 @@ const mappedChildrenArray4 = React.Children.map(mixedChildren, elementOrString =
 const mappedChildrenArray5 = React.Children.map(singlePluralChildren, element => element.key);
 // $ExpectType string[]
 const mappedChildrenArray6 = React.Children.map(renderPropsChildren, element => element.name);
+// $ExpectError
+const mappedChildrenArray7 = React.Children.map(nodeChildren, node => node).map;
```
Comment (OliverJAsh) [inline types/react/test/index.ts]: Previously this would not error.

**Review by eps1lon — COMMENTED** [review]
Would this work by sticking with two overloads instead using conditionals?

```ts
map<T, C extends null | undefined>(children: C | C[], fn: (child: C, index: number) => T): C;
map<T, C>(children: C | C[], fn: (child: C, index: number) => T): Array<Exclude<T, boolean | null | undefined>>;
```

**Comment by OliverJAsh** [comment]
@eps1lon I just tried it and it fails the new test (the one added in this PR). It seems to be using the second overload.

**Comment by eps1lon** [comment]
> @eps1lon I just tried it and it fails the new test (the one added in this PR). It seems to be using the second overload.

Well then lets stick with your version.

**Review by eps1lon — APPROVED** [review]

**Pull request merged**